### ISSUE TITLE
Potential bugs with random seeding

### ISSUE THREAD

**Issue by mooreryan** [bagnalla/ocamlzar#1]
Hi, I have run into a couple of potential bugs (or, they could definitely be my misunderstanding of what should be happening!)

- For `Zar.take`, the first value (end of the list) is always the same number (because it is always taking the `first` of the stream). This makes sense, but maybe not the most useful for sampling...eg you would need to do something like `die |> Zar.drop 1 |> Zar.take 10` or something to get a random sample.
- The random sampling seems to not respect the random seeding in a way that I would expect. If you look at the example program below, ...
  -  `test_it_1` function generate the first 14 values differently and then all for int lists will end with the same set of numbers.
  - `test_it_2` is also a bit weird in which gives two separate streams
  - Given that the source of randomness looks to be this line: `let rec bits () : bool stream = SCons (Random.bool (), bits)`, which uses Random module, I would expect re-initializing the Random module with the same seed will produce the same stream of random values, but that is not the case.

```ocaml
let sexp_of_int_list ints = 
  let open Sexplib.Sexp in
  List (List.map (fun i -> Atom (Int.to_string i)) ints)

let print_int_list ints = 
  ints |> sexp_of_int_list |> Sexplib.Sexp.to_string_mach |> print_endline

let test_it_1 () = 
  print_endline "test_it_1";
  Random.init 1234321; 
  let die = Zar.die 100 in
  die |> Zar.take 75 |> print_int_list;
  Random.init 1234321; 
  die |> Zar.take 75 |> print_int_list;
  Random.init 1234321; 
  let die = Zar.die 100 in
  die |> Zar.take 75 |> print_int_list;
  Random.init 1234321; 
  die |> Zar.take 75 |> print_int_list;
  ()

let test_it_2 () = 
  print_endline "test_it_2";
  Random.init 44553; 
  let die = Zar.die 100 in
  die |> Zar.take 75 |> print_int_list;
  Random.init 44553; 
  die |> Zar.take 75 |> print_int_list;
  Random.init 44553; 
  let die = Zar.die 100 in
  die |> Zar.take 75 |> print_int_list;
  Random.init 44553; 
  die |> Zar.take 75 |> print_int_list;
  ()


let () = 
  test_it_1 ();
  test_it_2 ();
  ()
```

Output 

```
test_it_1                        
(68 44 70 17 64 24 21 65 51 50 18 98 5 62 36 80 22 70 40 31 62 4 38 57 10 32 6 65 22 0 92 48 25 28 74 82 97 46 36 16 29 10 34 94 48 20 19 56 96 78 40 11 0 76 50 37 46 36 14 13 40 20 8 39 45 79 85 3 15 9 1 83 35 63 73)
(68 44 70 17 64 24 21 65 51 50 18 98 5 62 36 80 22 70 40 31 62 4 38 57 10 32 6 65 22 0 92 48 25 28 74 82 97 46 36 16 29 10 34 94 48 20 19 56 96 78 40 11 0 76 50 37 46 36 14 13 40 54 69 72 89 92 51 57 54 50 91 67 81 86 73)
(68 44 70 17 64 24 21 65 51 50 18 98 5 62 36 80 22 70 40 31 62 4 38 57 10 32 6 65 22 0 92 48 25 28 74 82 97 46 36 16 29 10 34 94 48 20 19 56 96 78 40 11 0 76 50 37 46 36 14 13 40 20 8 39 45 79 85 3 15 9 1 83 35 63 73)
(68 44 70 17 64 24 21 65 51 50 18 98 5 62 36 80 22 70 40 31 62 4 38 57 10 32 6 65 22 0 92 48 25 28 74 82 97 46 36 16 29 10 34 94 48 20 19 56 96 78 40 11 0 76 50 37 46 36 14 13 40 54 69 72 89 92 51 57 54 50 91 67 81 86 73)
test_it_2
(84 73 73 20 86 39 20 28 51 52 38 23 30 88 33 4 50 88 6 40 33 76 34 29 84 64 8 38 45 9 57 19 12 83 95 25 78 58 21 10 8 18 44 62 41 83 37 23 4 77 27 79 15 50 54 5 74 7 89 21 15 51 93 97 90 7 30 57 69 61 55 54 69 24 11)
(20 86 39 20 28 51 52 38 23 30 88 33 4 50 88 6 40 33 76 34 29 84 64 8 48 40 57 34 5 37 68 95 2 78 56 17 71 48 44 64 25 53 92 1 28 46 34 63 66 86 24 8 39 68 16 38 94 60 57 11 72 7 67 97 44 18 25 28 90 56 24 55 37 29 11)
(84 73 73 20 86 39 20 28 51 52 38 23 30 88 33 4 50 88 6 40 33 76 34 29 84 64 8 38 45 9 57 19 12 83 95 25 78 58 21 10 8 18 44 62 41 83 37 23 4 77 27 79 15 50 54 5 74 7 89 21 15 51 93 97 90 7 30 57 69 61 55 54 69 24 11)
(20 86 39 20 28 51 52 38 23 30 88 33 4 50 88 6 40 33 76 34 29 84 64 8 48 40 57 34 5 37 68 95 2 78 56 17 71 48 44 64 25 53 92 1 28 46 34 63 66 86 24 8 39 68 16 38 94 60 57 11 72 7 67 97 44 18 25 28 90 56 24 55 37 29 11)
```

**Comment by bagnalla** [comment]
Hi Ryan,

> For Zar.take, the first value (end of the list) is always the same number

Would you expect it to consume the samples from the stream so that subsequent calls to `take` produce new samples? Here's one way I could imagine supporting that:
1) Provide a 'takedrop' function (perhaps called something else) that implements what you suggest. Though, this would still not modify the input stream but would return a tuple containing the list of samples and the rest of the stream. And,
2) provide a stateful interface that internally keeps a mutable reference to a sample stream and uses 'takedrop' to implement the desired behavior. Maybe something like:
```OCaml
class ['a] sampler : 'a stream -> object
  method gen : int -> 'a list
end
```
in stream.mli, and
```OCaml
class ['a] sampler (init : 'a stream) = object
  val mutable state = init
  method gen (n : int) : 'a list =
    let samples, state' = takedrop n state in
    state <- state';
    samples
end
```
in stream.ml.

Let me know if I've misunderstood your point.

The stream interface was meant to be somewhat temporary until we figured out something better, so I'm not surprised to see some warts in its usability, and I greatly appreciate feedback like this. Another wrinkle (that you seem to have noticed) is that 'take' reverses the order of the elements taken from the stream, which technically shouldn't really matter for random samples, but might be inconvenient for testing purposes. I'll look into changing that -- If you have any thoughts on the matter I would be glad to hear them.

I'll look into your second point more closely later.

-- Alex

**Comment by mooreryan** [comment]
Regarding the stream interface--I have a fork in which I converted the stream to use OCaml standard libraries `Seq.t` ...it was a pretty straightforward change as that interface is quite similar to the stream interface you have currently. Would you be interested in that sort of change? I did not do any performance testing of that however.

Regarding the `take` behavior, I will get back to you on that later today. (Let me see if I can make a better example to show.)

**Comment by bagnalla** [comment]
I would definitely be open to eliminating the stream type and using `Seq.t` instead.

It might also be nice to codify some of our expectations of the provided interface's behavior as test cases in `test/zar.ml`.

**Comment by mooreryan** [comment]
I can open draft pull requests with some of this stuff later today or tomorrow for you to look at. It will be simpler to discuss concrete code and quickcheck tests I think!

**Comment by bagnalla** [comment]
I'm not quite sure how to explain the behavior in `test_it_1` and `test_it_2`,  but I'm thinking that the streams of bits produced by `bits` really ought to be memoized (e.g., via `Seq.memoize`), so that not only the first element will be the same across the two calls to `Zar.take` but all of the elements will be.

It was never my intention that a single bit stream would produce different random bits when viewed multiple times -- for that you should have to create multiple separate bit streams. E.g., if the first five bits of a stream are "01001", then they will always be "01001", and if you want more random bits you'll need to go further into the stream or create a new one.

I think memoizing would solve both issues you pointed out (although I'm not 100% sure), but then we might want to provide a more convenient way to consume streams besides calling `take` and `drop` separately (as I was discussing above).

In any event, I think it would be wise to convert to `Seq.t`. If you want to do this then I would gladly accept a PR, otherwise I'll take a stab at it some time over the next few days.

-- Alex

**Comment by mooreryan** [comment]
> It was never my intention that a single bit stream would produce different random bits when viewed multiple times -- for that you should have to create multiple separate bit streams. E.g., if the first five bits of a stream are "01001", then they will always be "01001", and if you want more random bits you'll need to go further into the stream or create a new one.

It is interesting, but that behavior (where you get different streams with repeated calls to take) is how it is working currently.  You can see that it is so in the quickcheck tests in that draft pull request.

By the way...I think the problems in the test 1 and test 2 above are at least partly caused by the reseeding of the random number generator is not done at exactly the correct times. That quickcheck code I added seeds the generator at the correct moments to results that make more sense.

Regarding the caching...you may want to take a look at the persistent iterators from [iter package](https://c-cube.github.io/iter/1.4/iter/Iter/index.html#caching). That's a different type of iterator than Seq is but maybe that is a good model for the caching anyway (if that's what you decide to pursue).

Regarding the using of `Seq.t` it is actually interesting but I'm getting a slightly different behavior when doing what seems like a straightforward conversion to Seq.t. I will look at it for a while longer, and open a draft pull request so you can see what I mean.

**Comment by bagnalla** [comment]
Fixed in ef83e014efbdd89ea0464f693c6fab9c129a2d73.
